napari/_qt/containers/layers/_layer_model.py

```diff
@@ -0,0 +1,82 @@
+from typing import TYPE_CHECKING
+
+from qtpy.QtCore import QModelIndex, QSize, Qt
+from qtpy.QtGui import QImage
+from qtpy.QtWidgets import QWidget
+
+from ....layers import Layer
+from ..list import QtListModel
+
+if TYPE_CHECKING:
+
+    from ....components.layerlist import LayerList
+
+
+class QtLayerListModel(QtListModel[Layer]):
+    LayerRole = Qt.UserRole
```

Comment:
I'll be honest, I'm not so familiar with the `UserRole` stuff in Qt .... will need to look up!!

Reply:
the general idea of data roles is nice (and simple really).  in a `QAbstractItemModel`, there is a `data(self, index: QModelIndex, role: Qt.ItemDataRole)` method that lets you ask for some data associated with a particular item in the model/list/tree.  the "role" can be something like "display role" (like layer.name), or "checkStateRole" (which I'm using for layer.visible)... but `role` is just an integer enum, and you can add your own  custom roles.  They offer `UserRole` as a high integer enum and I'm using it here to just get the actual item at the index.  The main thing people might do when subclassing these model/view classes is to define their own `data` and `setData` methods, with custom logic for the specific type of object in their list.